### DISCUSSION TITLE
How to create interactions like ncdu, htop, and mc do?

### DISCUSSION THREAD

**Discussion by Irineu333** [ajalt/clikt#478] [General]
I'm new to the world of terminal programs, and these particular programs amaze me a lot, they are different from the usual, I want to understand how to make my program react to keyboard interaction and still have features like scrolling and even recognize mouse clicks.

**Comment by ajalt** [comment]
On JVM, [JLine][1] can handle mouse clicks. Then you can use something like [mordant][2] or [mosaic][3] to generate the UI.

[1]: https://github.com/jline/jline3
[2]: https://github.com/ajalt/mordant
[3]: https://github.com/JakeWharton/mosaic

**Reply by Irineu333** [reply]
The idea of Mosaic is brilliant, using the runtime of Compose to create a reactive UI. It has a lot of potential, thank you for introducing it to me.

**Reply by Irineu333** [reply]
I'm already using Mordant. I'll take a look at JLine, thank you.

**Reply by Irineu333** [reply]
I think JLine might suit my goals. In combination with mordant, which is more mature. Interaction with mouse events is optional. What interested me in JLine was the function to enter raw mode, and this allows for directly handling the pressed keys. Thank you very much!

It would be really cool if mordant had it, but it seems to require native binaries, and I believe you want to keep your library 100% Kotlin.
